File: app/javascript/channels/messages_channel.js
import consumer from "./consumer"

const messagesChannel = {
  create: (roomNumber, receivedFun) => {
    return consumer.subscriptions.create({channel: "MessagesChannel", room_number: roomNumber}, {
      initialized() {
        this.sendMessage = this.sendMessage.bind(this)
      },

      received(data) {
        receivedFun.call(this, data.html)
      },

      sendMessage(message) {
        this.send({message_text: message})
      },
    });
  }
}

export default messagesChannel

File: app/channels/messages_channel.rb
class MessagesChannel < ApplicationCable::Channel
  def subscribed
    @room_number = params[:room_number]
    stream_from broadcast_name
  end

  def receive(data)
    message = Message.new(data['message_text'], current_user)
    message_html = ApplicationController.render(partial: 'messages/message', locals: { message: message })
    ActionCable.server.broadcast(broadcast_name, html: message_html)
  end

  private

  def broadcast_name
    format('room_%s::messages', @room_number)
  end
end
